File: src/docforge/openapi/refs.py
_SCHEMA_PREFIX = "#/components/schemas/"
_FROM_KINDS = {"endpoint", "schema"}


def ref_to_schema_name(ref: str) -> str | None:
    if not isinstance(ref, str) or not ref.startswith(_SCHEMA_PREFIX):
        return None
    name = ref[len(_SCHEMA_PREFIX):]
    return name or None


def ref_link(ref: str, *, from_kind: str) -> tuple[str, str]:
    if from_kind not in _FROM_KINDS:
        raise ValueError(f"from_kind must be one of {_FROM_KINDS}, got {from_kind!r}")
    name = ref_to_schema_name(ref)
    if name is None:
        return ref, ref
    if from_kind == "endpoint":
        return name, f"../schemas/{name}.md"
    return name, f"{name}.md"

File: src/openapi/refs.ts
const SCHEMA_PREFIX = "#/components/schemas/";
const FROM_KINDS = new Set(["endpoint", "schema"]);

export type FromKind = "endpoint" | "schema";

export function refToSchemaName(ref: unknown): string | null {
  if (typeof ref !== "string" || !ref.startsWith(SCHEMA_PREFIX)) return null;
  const name = ref.slice(SCHEMA_PREFIX.length);
  return name || null;
}

export function refLink(
  ref: string,
  opts: { fromKind: FromKind },
): [string, string] {
  if (!FROM_KINDS.has(opts.fromKind)) {
    throw new Error(`fromKind must be 'endpoint' or 'schema', got '${opts.fromKind}'`);
  }
  const name = refToSchemaName(ref);
  if (name === null) return [ref, ref];
  if (opts.fromKind === "endpoint") return [name, `../schemas/${name}.md`];
  return [name, `${name}.md`];
}
